src/UIAutomation/UiaOperationAbstraction/UiaOperationAbstraction.h

```diff
@@ -493,6 +510,33 @@ namespace UiaOperationAbstraction
 
             return result;
         }
+
+        // The helper allows "normalizing" the given object to a default-constructed instance of `T` if the
+        // provided `shared_ptr` is empty.
+        //
+        // This is useful for creating abstraction types whose nested types -- when default-constructed --
+        // may not have valid values.
+        //
+        // For instance, `UiaTuple<UiaArray<UiaInt>>` is a combination of:
+        // 1. `UiaTuple<T>` = `std::shared_ptr<std::tuple<T>>`
+        // 2. `UiaArray<T>` = `std::shared_ptr<std::vector<T>>`
+        // 3. `UiaInt` = `int`
+        //
+        // If the constructor of `UiaTuple<T>` simply calls `std::make_shared<T>` to create the local value,
+        // the inner value of `T` will be default-constructed.
+        //
+        // In the above case, `std::shared_ptr<std::vector<T>>`'s default construction is about creating an
+        // empty `std::shared_ptr` which is an unexpected/uninitialized value from the perspective of the
+        // abstraction library tools and therefore encountering it means exceptions.
+        //
+        // The helper allows default-constructed wrapper type objects to be correctly constructed even if
+        // their corresponding local types require more work than calling their default constructors to be
+        // correctly initialized (here, the local types are `std::shared_ptr`-based).
+        template<class T>
+        std::shared_ptr<T> GetOrMakeSharedPtr(const std::shared_ptr<T>& ptr)

```

Comment:
This is such an awesome helper!

Reply:
Glad you like it! :)

---
In reply to: [504926806](https://github.com/microsoft/Microsoft-UI-UIAutomation/pull/43#discussion_r504926806) [](ancestors = 504926806)